src/v/utils/retry_chain_node.h

```diff
@@ -154,6 +154,211 @@
 #include <ranges>
 #include <variant>
 
+namespace detail {
+/// A fixed-size circular buffer that behaves like std::vector for basic
+/// operations. When the buffer reaches its maximum capacity, new elements
+/// overwrite the oldest elements in FIFO order.
+template<typename T>
+class circular_buffer {
+public:
+    using value_type = T;
+    using reference = T&;
+    using const_reference = const T&;
+
+    /// Create a circular buffer with the specified maximum capacity
+    explicit circular_buffer(size_t capacity)
+      : _data()
+      , _capacity(capacity) {}
+
+    /// Add an element to the buffer. When the buffer is full, the oldest
+    /// element is overwritten.
+    void push_back(const T& value) {
+        if (_size < _capacity) {
+            _data.push_back(value);
+            ++_size;
+        } else {
+            _data[_start] = value;
+            _start = (_start + 1) % _capacity;
+        }
+        _total_added++;
+    }
+
+    void push_back(T&& value) {
+        if (_size < _capacity) {
+            _data.push_back(std::move(value));
+            ++_size;
+        } else {
+            _data[_start] = std::move(value);
+            _start = (_start + 1) % _capacity;
+        }
+        _total_added++;
+    }
+
+    void push_back(std::string_view sv) {
```

Comment:
feels a bit odd having this interface bound to `basic_string_view<char>` when the container is generic. could remove the top-level template and call out that it's for signed chars only, or restrict the template to char types and have this function take `basic_string_view<T>`, or have this function take a generic range.

Reply:
fixed